### PR TITLE
Visual search update

### PR DESCRIPTION
**Pull request by jn2clark**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
It introduces:
1) additional heuristic methods for the image preprocessing via chunking (or patching) (see https://marqo.pages.dev/Preprocessing/Images/)
2) additional models for the image preprocessing via chunking (or patching) (see https://marqo.pages.dev/Preprocessing/Images/)
3) a text-image based reranker that allows reranking of images based on an open-vocabulary model (owl-vit) 
4) An example script and streamlit app to demonstrate the new functionality
5) tests for the new functionality

* **What is the current behavior?** (You can also link to an open issue here)
1) Currently we have two methods for image based pre-processing. A simple grid based heuristic and a faster-rcnn based model from pytorch (trained on mscoco)
2) There is no option for open-vocabulary localization or re-ranking based on images

* **What is the new behavior (if this is a feature change)?**
1) The new behavior is there are more options to select a model for pre-proessing. A yolox based onnx model trained on LVIS is added, models based on determining salient regions of images from attention maps is included (based on DINO)
2) There is a new heuristic based model which uses overlapping boxes instead of a non-overlapping grid.
3) Re-ranking of images from text based queries is possible

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
Yes

* **Related Python client changes** (link commit/PR here)
None

* **Related documentation changes** (link commit/PR here)
To be included

* **Other information**:
api tests are here 
https://github.com/marqo-ai/marqo-api-tests/tree/visual_search_update
Manual testing end-to-end via the demo has been performed as well

* **Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by pandu-k** [comment]
For the [advanced chunking documentation](https://marqo.pages.dev/Preprocessing/Images/#advanced-chunking) can we explain the implications of advanced chunking? Like this:

"... With advanced chunking, highlights will be more likely to correspond directly with human recognisable objects, such as a hand or face. See here(insert link) for more info"

Also can we make the language consistent: chunking XOR patching, or be explain the difference between the terms?

File: examples/ImageSearchLocalization/app.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    from PIL import Image, ImageDraw
+        
+    # create rectangle image
+    img1 = ImageDraw.Draw(img)  
+    img1.rectangle(box, outline  = 'orange', width = 8)
+    return img
+
+def main():
+
+    data_dir = "/home/jesse/code/stable-diffusion/scripts/coco/"
```
Comment (pandu-k) [inline examples/ImageSearchLocalization/app.py:77]: Probably change this to a placeholder, and add instructions (or a link to instructions) on how to set up the data dir

File: examples/ImageSearchLocalization/app.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    image_location = 'image_location'
+    highlights = '_highlights'
+    device = 'cuda'
+    
+    docker_prefix = 'http://host.docker.internal:8222/'
+    locale_location = '/home/jesse/code/stable-diffusion/scripts/coco/'
```
Comment (pandu-k) [inline examples/ImageSearchLocalization/app.py:83]: Change this to a placeholder, and add instructions (or a link to instructions) on how to set up

File: examples/ImageSearchLocalization/index_all_data.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+local_dir = os.getcwd() + '/images/'
+if not use_remote:
+    zip_file_url = 'https://marqo-public-datasets.s3.us-east-2.amazonaws.com/demos/ImageSearchLocalisation/images.zip'
+    zip_file = download_file(zip_file_url, local_dir=local_dir)
+    extract_zip(zip_file, local_dir=local_dir)
+    # alternatively you can download the files individually
+    #downloaded = download_parallel(urls=data['s3_uri'].tolist(), local_dir=local_dir)
+    locators = glob.glob(local_dir + '*.jpg')
+    docker_path = 'http://host.docker.internal:8222/'
```
Comment (pandu-k) [inline examples/ImageSearchLocalization/index_all_data.py:36]: Is this better named something like `image_server_docker_path`?

File: examples/ImageSearchLocalization/utils.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        full_local_path, _ = urlretrieve(url, full_local_path)
+
+    return full_local_path
+
+def download_parallel(urls: List[str], local_dir: str, n_processes=8) -> List[str]:
+    
+    N = len(urls)
+    print(f"downloading {N} urls to {local_dir} using {n_processes} processes")
+
+    func = partial(download_files, local_dir=local_dir)
```
Comment (pandu-k) [inline examples/ImageSearchLocalization/utils.py:45]: nit: can we use a more descriptive name? Like `download_to_dir`

File: examples/ImageSearchLocalization/utils.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    urls_split = np.array_split(urls, n_processes)
+    urls_split = [split.tolist() for split in urls_split]
+    with mp.Pool(n_processes) as pool:
+        results = pool.map(func, urls_split)
+
```
Comment (pandu-k) [inline examples/ImageSearchLocalization/utils.py:51]: nit: multithreading can probably achieve similar results with less overhead as downloading is IO bound

File: examples/ImageSearchLocalization/app.py (hunk 1 above)
Comment (jn2clark) [reply examples/ImageSearchLocalization/app.py:77]: Thanks. Yeah, the demo is still a WIP

File: src/marqo/s2_inference/processing/DINO_utils.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    model.eval()
+    model.to(device)
+
+    return model, _get_DINO_transform(image_size=image_size)
+
+def _get_DINO_transform(image_size: Tuple = (224, 224)) -> Any:
+    """gets the preprocessing transform for dino models
+
+    Args:
+        image_size (Tuple, optional): _description_. Defaults to (224, 224).
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/DINO_utils.py:67]: nit: describing the default value is redundant if it is also declared in the function signature.

Also, removing the placeholder `_description_` will make the doc string neater.

File: src/marqo/s2_inference/processing/DINO_utils.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    Returns:
+        Any: _description_
+    """
+    return pth_transforms.Compose([
+    pth_transforms.Resize(image_size),
+    pth_transforms.ToTensor(),
+    pth_transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/DINO_utils.py:75]: Can we be explicit about where these floats come from? This helps code readability.

File: src/marqo/s2_inference/processing/DINO_utils.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    ])
+
+def DINO_inference(model: Any, transform: Any, img: ImageType = None, 
+                        patch_size: int = None, device: str = "cpu") -> FloatTensor:
+    """runs inference for a model, transform and image
+
+    Args:
+        model (Any): ('vit_small', 'vit_base')
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/DINO_utils.py:83]: nit: consider using the literal type for options: https://stackoverflow.com/a/58521635 (I just read about this today)

File: src/marqo/s2_inference/processing/DINO_utils.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        device (str, optional): _description_. Defaults to "cpu".
+
+    Returns:
+        FloatTensor: _description_
+    """
+    
+    img = transform(img)
+
+    # make the image divisible by the patch size
+    w, h = img.shape[1] - img.shape[1] % patch_size, img.shape[2] - img.shape[2] % patch_size
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/DINO_utils.py:96]: Would `w, h == 0, 0`? or am I missing something?

File: src/marqo/s2_inference/processing/DINO_utils.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    return model, _get_DINO_transform(image_size=image_size)
+
+def _get_DINO_transform(image_size: Tuple = (224, 224)) -> Any:
+    """gets the preprocessing transform for dino models
+
+    Args:
+        image_size (Tuple, optional): _description_. Defaults to (224, 224).
+
+    Returns:
+        Any: _description_
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/DINO_utils.py:70]: Can we document the returned object?

File: src/marqo/s2_inference/processing/DINO_utils.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    Args:
+        model (Any): ('vit_small', 'vit_base')
+        transform (Any): _get_DINO_transform
+        img (ImageType, optional): _description_. Defaults to None.
+        patch_size (int, optional): _description_. Defaults to None.
+        device (str, optional): _description_. Defaults to "cpu".
+
+    Returns:
+        FloatTensor: _description_
+    """
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/DINO_utils.py:91]: Can we document return dimensionality?

File: src/marqo/s2_inference/processing/image.py [hunk 12]
```diff
@@ -1,300 +1,141 @@
...
+        patch.process()
+    except PIL.UnidentifiedImageError as e:
+        raise ChunkerError from e
 
+    return patch.patches,patch.bboxes_orig
 
 
 class PatchifySimple:
-    """class to do the patching
+    """class to do the patching. this one creates non-pverlapping boixes and chunks the image
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/image.py:117]: Typo in `non-pverlapping `

File: src/marqo/s2_inference/processing/image.py [hunk 13]
```diff
@@ -306,143 +147,279 @@ def process(self):
...
+        Args:
+            device (str, optional): _description_. Defaults to 'cpu'.
+            size (Tuple, optional): _description_. Defaults to (224, 224).
+            min_area (float, optional): _description_. Defaults to 60*60.
+            nms (bool, optional): _description_. Defaults to True.
+            replace_small (bool, optional): _description_. Defaults to True.
+            top_k (int, optional): _description_. Defaults to 10.
+            filter_bb (bool, optional): _description_. Defaults to True.
+            min_area_replace (float, optional): _description_. Defaults to 60*60.
+        """
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/image.py:167]: Would be more readable if each arg's role is explained (how each arg affects the function's behaviour)

File: src/marqo/s2_inference/processing/image_utils.py [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+    Args:
+        image (ImageType): _description_
+        bboxes (Union[List[float], FloatTensor, ndarray]): _description_
+
+    Returns:
+        List[ImageType]: _description_
+    """
+    return [image.crop(bb) for bb in bboxes]
+
+def _process_patch_method(method: str) -> Tuple[str, dict]:
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/image_utils.py:278]: Is this function used to process incoming calls to Marqo?

File: src/marqo/s2_inference/processing/utils.py [hunk 15]
```diff
@@ -0,0 +1,834 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (pandu-k) [inline src/marqo/s2_inference/processing/utils.py:1]: If most of these utils come from an OSS package - maybe rename it? For example "oss_utils.py"

File: src/marqo/s2_inference/reranking/model_utils.py [hunk 16]
```diff
@@ -346,16 +351,17 @@ def _predict_owl(model, processed_inputs, post_process_function, size):
...
         target_sizes = torch.Tensor([size[::-1]])
         # Convert outputs (bounding boxes and class logits) to COCO API
         results = post_process_function(outputs=outputs, target_sizes=target_sizes)
-
         return results
 
-def process_owl_results(results: Dict) -> List:
-    """_summary_
+def process_owl_results(results: List) -> List:
+    """wrapper for processing a list of resultsd from owl-vitexi
```
Comment (pandu-k) [inline src/marqo/s2_inference/reranking/model_utils.py:364]: typo in `resultsd`

Comment (pandu-k) [inline src/marqo/s2_inference/reranking/model_utils.py:364]: typo in `resultsd`

File: src/marqo/s2_inference/reranking/rerank.py [hunk 17]
```diff
@@ -18,24 +23,60 @@ def rerank_search_results(search_result: Dict, query: str, model_name: str, devi
...
-        reranker = ReRankerOwl(model_name=model_name, device=device, image_size=(240,240))
-        reranker.rerank(query=query, results=search_result, image_attributes=searchable_attributes)
+        # owl needs the image location, while the text based ones can handle different number of fields but concat the text
+        if searchable_attributes in (None, [], (), ''):
+            raise RerankerError(f"found searchable_attributes={searchable_attributes} but expected list of strings for {model_name}")
+        
+        if len(searchable_attributes) > 1:
+            logger.info(f"currently only a single attribute can be reranked over for {model_name}. taking the first field {[searchable_attributes[0]]} from {searchable_attributes}")
+            searchable_attributes = [searchable_attributes[0]]
+
```
Comment (pandu-k) [inline src/marqo/s2_inference/reranking/rerank.py:39]: This needs to be communicated to users. For example:
- a warning accompanying a multiple fields rerank request
- an error if they attempt over multiple fields
- The API specs prevent it

File: src/marqo/tensor_search/tensor_search.py [hunk 18]
```diff
@@ -762,10 +762,15 @@ def search(config: Config, index_name: str, text: str, result_count: int = 3, hi
         )
     else:
         raise errors.InvalidArgError(f"Search called with unknown search method: {search_method}")
+
+    logger.info("reranking using {}".format(reranker))
```
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py:778]: Wouldn't this best be output only when reranker is not None?

File: src/marqo/s2_inference/processing/DINO_utils.py (hunk 9 above)
Comment (jn2clark) [reply src/marqo/s2_inference/processing/DINO_utils.py:96]: i think it is the order of operations.

**Comment by jn2clark** [comment]
@pandu-k closing this and opening an updated one after remerging from mainline. https://github.com/marqo-ai/marqo/pull/214

**Pull request closed without merging**